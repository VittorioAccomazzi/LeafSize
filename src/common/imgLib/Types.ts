
/**
 * definition of a colour. Each channel is considered an 8 bit integer.
 * the values are *not* pre multiplied by alpha. If alpha is not present 
 * it is considered 255, and so the colour is fully opaque.
 */
export interface Colour {
    r : number,
    g : number,
    b : number,
    a?: number
}

/**
 * 2D point. Defined here because DOMPoint requires many more parameters
 * which are not necessary for the imgLib purposes.
 */
export interface Point {
    x: number,
    y: number
}

/**
 * generic interface for an image/mask object. Something which can be displayed and/or overlayed on an image.
 */
export default interface IImage<T> {
    readonly width : number,
    readonly height: number,
    set : (x:number, y:number, v : T )=> void,
    get : (x:number, y:number) => T,
    foreachPixel : (func:(x:number, y:number, v:T)=>void)=>void
}