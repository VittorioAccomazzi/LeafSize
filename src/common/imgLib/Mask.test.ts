import Mask, { isMask } from './Mask';
import {Point} from './Types';
import {ImageFactory, isImage} from './ImageBase';
import CanvasUtils from '../utils/CanvasUtils';
import hash, { loadCanvas, testImage } from '../utils/TestUtils';
import ColourModels from './ColourModels';


describe('Mask', ()=>{

    test('shall create an empty mask', ()=>{
        const width = 8;
        const height= 5;
        const mask = new Mask(width, height);
        expect(mask.width).toBe(width);
        expect(mask.height).toBe(height);
        mask.foreachPixel((x,y,v) => expect(v).toBe(false));
    })

    test('shall allow to set and read the values', ()=>{
        const mask = new Mask(8,5);
        const x = 3;
        const y = 2;
        mask.set(x,y,true);
        expect(mask.get(x,y)).toBe(true)

        expect(()=>{
            mask.set(88,2,false);
        }).toThrow();

        expect(()=>{
            mask.get(88,2);
        }).toThrow();
    })

    test('shall calculate the area',()=>{
        const mask = new Mask(8,5);

        mask.set(2,3,true);
        expect(mask.area).toBe(1)

        mask.set(1,3,true);
        expect(mask.area).toBe(2)

        mask.set(1,1,true);
        expect(mask.area).toBe(3)
    })

    test('shall allow to access the pixel array',()=>{
        const width =8;
        const height=5;
        const mask = new Mask(width, height);
        mask.set(2,3,true);
        mask.set(3,1,true);
        mask.set(3,3,true);
        const pixels = mask.imagePixels;
        mask.foreachPixel((x,y,v) => {
            const offset = x+y*width;
            expect(pixels[offset]).toBe(v);
        });
    })

    test('shall look thought all the pixel of the mask', ()=>{
        let nPixels = 0;
        let xMax = 0;
        let yMax = 0;
        const width =8;
        const height=5;
        const mask = new Mask(width, height);
        mask.foreachPixel((x,y,v) => {
            nPixels++;
            xMax = Math.max(x, xMax);
            yMax = Math.max(y, yMax)
        });
        expect( nPixels ).toBe( width*height);
        expect( xMax ).toBe( width-1 );
        expect( yMax ).toBe( height-1 );
    })

    test('shall invert the mask', ()=>{
        const width =8;
        const height=5;
        const mask = new Mask(width, height);
        const xP=3;
        const yP=2;
        mask.set(xP,yP,true);
        mask.Not()
        mask.foreachPixel((x,y,v)=>{
            if( x === xP && y === yP) {
                expect(v).toBe(false);
            } else {
                expect(v).toBe(true);
            }
        })
    })

    test('shall AND two masks',()=>{
        const width =8;
        const height=5;
        const xP=3;
        const yP=2;
        const mask1 = new Mask(width, height);
        const mask2 = new Mask(width, height);

        mask1.set(xP,yP,true);
        mask2.set(xP,yP,true);
        mask1.set(1,1,true);
        mask2.set(2,2,true);

        mask1.And(mask2);

        mask1.foreachPixel((x,y,v)=>{
            if( x === xP && y === yP) {
                expect(v).toBe(true);
            } else {
                expect(v).toBe(false);
            }
        })
    })

    test('shall OR two masks', ()=>{
        const width =8;
        const height=5;
        const mask1 = new Mask(width, height);
        const mask2 = new Mask(width, height);

        const pt1 = [[1,2],[3,2]];
        const pt2 = [[1,2],[2,3]];

        pt1.forEach(v=>mask1.set(v[0],v[1],true));
        pt2.forEach(v=>mask2.set(v[0],v[1],true));

        mask1.Or(mask2);

        mask1.foreachPixel((x,y,v)=>{
            const inP1 = pt1.find(v=>v[0]===x&&v[1]===y);
            const inP2 = pt2.find(v=>v[0]===x&&v[1]===y);
            if( inP1 || inP2 ){
                expect(v).toBe(true)
            } else {
                expect(v).toBe(false)
            }
        })
    })

    test('Shall clone the mask',()=>{
        const width =8;
        const height=5;
        const xP=3;
        const yP=2;
        const mask1 = new Mask(width, height);
        mask1.set(xP, yP, true);
        const mask2 = Mask.Clone(mask1);

        // make sure they are identical
        mask1.foreachPixel((x,y,v)=>{
            expect(mask2.get(x,y)).toBe(v);
        })

        // it shall be a shallow copy
        mask1.set(1,2,true);
        mask1.set(2,2,true);
        mask2.foreachPixel((x,y,v)=>{
            expect(v).toBe(x === xP && y === yP);
        })
    })

    test('shall provide minus operator', ()=>{
        const width =8;
        const height=5;
        const mask1 = new Mask(width, height);
        const mask2 = new Mask(width, height);

        mask1.set(2,3,true);
        mask2.set(2,3,true);

        const xP =3;
        const yP =2;

        mask1.set(xP,yP, true);

        mask1.Minus(mask2);

        mask1.foreachPixel((x,y,v)=>{
            expect(v).toBe( x===xP && y===yP);
        })
    })

    test('shall provide fill operation ', ()=>{
        const width = 128;
        const height= 128;
        const num = 500;

        const NeigbourdPoints : Point [] = [ { x:-1, y:0 },
            { x:1, y:0 },
            { x:0, y:1 },
            { x:0, y:-1 },
            { x:1, y:1 },
            { x:-1, y:1 },
            { x:1, y:-1 },
            { x:-1, y:-1 }
        ]

        const mask = new Mask(width, height);

        const xSeed = width/4;
        const ySeed = height/2

        let x = xSeed;
        let y = ySeed;

        // generate a random path between 0 to width-10
        for(let i=0; i<num; i++ ){
            mask.set(x,y,true);
            const index = Math.floor( Math.random() * NeigbourdPoints.length );
            const pt = NeigbourdPoints[index];
            x += pt.x;
            y += pt.y;
            
            x = Math.min(x, width-10-1);
            x = Math.max(x, 0 );
            y = Math.min(y, height-1);
            y = Math.max(y, 0)
        }

        // add two extra point in the are not covered by the path
        mask.set(width-5, height/2, true);
        mask.set(width-5, height/2+1, true);

        // get connected components.
        const cMask = mask.Fill(xSeed, ySeed);

        expect(cMask.area).toBe(mask.area-2);

        cMask.foreachPixel((x,y,v)=>{
            if( x < width-10 ){
                expect(v).toBe(mask.get(x,y))
            } else {
                expect(v).toBe(false);
            }
        })
    })

    test('shall provide the dilate operation', ()=>{
        const width = 128;
        const height= 128;
        const num = 500;
        const mask = new Mask(width,height);

        for(let n=0; n<num; n++ ){
            const x = Math.floor(Math.random()*width);
            const y = Math.floor(Math.random()*height);
            mask.set(x,y,true);
        }
        mask.set(10,10,true);

        const halfMask = 1;
        const dMask = mask.Dilate(halfMask);
        expect(dMask.area).toBeGreaterThan(mask.area);

        dMask.foreachPixel((x,y,v)=>{
                if( x>= halfMask && x<width-halfMask && y>=halfMask && y<height-halfMask ){
                // there must be a pixel set to true in the mask
                const xMin = Math.max(0, x-halfMask);
                const xMax = Math.min(width, x+halfMask+1);
                const yMin = Math.max(0, y-halfMask);
                const yMax = Math.min(height, y+halfMask+1);
                let val = false;
                for( let yK = yMin; yK<yMax; yK++){
                    for(let xK=xMin; xK<xMax; xK++) {
                        val ||= mask.get(xK,yK);
                    }
                }
                expect(v).toBe(val);
            }
        })
    })

    test('shall provide instance of',()=>{
        const myMask = new Mask(4,8);
        const myImage= ImageFactory.Uint8(4,8);

        expect (isMask(myMask)).toBeTruthy();
        expect (isMask(myImage)).toBeFalsy();
        expect (isImage(myMask)).toBeFalsy();
        expect (isImage(myImage)).toBeTruthy();
    })

    test('shall overlay the mask on the image', async ()=>{
        const canvas = await loadCanvas(testImage,8);
        const imgData= CanvasUtils.GetImageData(canvas);
        const {hue, sat } = ColourModels.toHSV(imgData);
        const mask = new Mask(sat.width, sat.height);
        const maskPixels = mask.imagePixels;
        const thr = 80/255;
        sat.imagePixels.forEach((v,i)=>{ maskPixels[i] = v < thr; })
        const bMask = mask.Fill(1,1); // hole filling
        bMask.Overlay(imgData,{r:0, g:0, b:0, a:128 });
        CanvasUtils.PutImageData(canvas,imgData);
        const hsh1 = await hash(canvas,'overlay');
        expect(hsh1).toMatchSnapshot();
        const line = bMask.Dilate(1);
        line.Minus(bMask);
        line.Overlay(imgData,{r:255,g:255,b:0});
        CanvasUtils.PutImageData(canvas, imgData);
        const hsh2 = await hash(canvas,'overlay-line');
        expect(hsh2).toMatchSnapshot();
    })

    test('shall return the connected components',async()=>{
        const canvas = await loadCanvas(testImage,12);
        const imgData= CanvasUtils.GetImageData(canvas);
        const {hue, sat } = ColourModels.toHSV(imgData);
        const mask = new Mask(sat.width, sat.height);
        const maskPixels = mask.imagePixels;
        const thr = 80/255;
        sat.imagePixels.forEach((v,i)=>{ maskPixels[i] = v < thr; })
        const bMask = mask.Fill(1,1); // hole filling
        const dMask = bMask.Dilate(1); // split small bridges
        dMask.Not(); // foreground components
        const masks = dMask.Components(8);
        // draw on a mask
        const colours = [ {r:255,g:255,b:0}, {r:255,g:0,b:0}, {r:0,g:255,b:0}, {r:0,g:255,b:255}, 
                          {r:255,g:255,b:255}, {r:255,g:0,b:255}, {r:0,g:0,b:255}, {r:0,g:0,b:0}];
        masks.forEach((m,i)=>{
            const d = m.Dilate(1);
            d.Minus(m)
            d.Overlay(imgData,colours[i])
        })
        CanvasUtils.PutImageData(canvas,imgData);
        const hsh = await hash(canvas,'components')
        expect(hsh).toMatchSnapshot();
    })

    test('Shall determine the bounding box', async()=>{
        const canvas = await loadCanvas(testImage,12);
        const imgData= CanvasUtils.GetImageData(canvas);
        const {hue, sat } = ColourModels.toHSV(imgData);
        const mask = new Mask(sat.width, sat.height);
        const maskPixels = mask.imagePixels;
        const thr = 90/255;
        sat.imagePixels.forEach((v,i)=>{ maskPixels[i] = v < thr; })
        const dMask = mask.Dilate(1);
        dMask.Not(); 
        dMask.Overlay(imgData,{r:255,g:255,b:0,a:255});
        CanvasUtils.PutImageData(canvas,imgData);
        const bbx = dMask.Boundaries();
        const ctx = canvas.getContext('2d');
        ctx.strokeStyle = "#FF0000";
        ctx.beginPath()
        ctx.rect(bbx.ulc.x, bbx.ulc.y, bbx.lrc.x-bbx.ulc.x, bbx.lrc.y-bbx.ulc.y);
        ctx.stroke();
        const hsh = await hash(canvas,'bounding box');
        expect(hsh).toMatchSnapshot();
    })
})

