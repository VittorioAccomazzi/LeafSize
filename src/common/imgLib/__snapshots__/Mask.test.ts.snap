// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mask Shall determine the bounding box 1`] = `"ffffffffffffffe00000000000000701ffffffffffffffe00000000000000f03fffffffffffffff80000000000000f03fffffffffffffff80000000000000f03ffffffffe3fffff80000000000000703fffffffffcfffffc0000000000000703fffffffffe7ffff80000000000000703ffffffffffbffffc0000000000000503ffffffffffdffffc0000000000000f0bffffffffffcffff80000000000020f0bffffffffffe7fff8000000000001070bfffffffffff3bff8000000000001070bfffffffffff99ffc0000000000000303fffffffffffc5ffc000000000000830bffffe00000000000000000000000530bffffe600000000000000000000007b1bffffef21fe1027fc0000000006007f1bffffee70fe7803f8000000800e003f3fffffecf87c7803f8000000c00f003f0fffffedfc78fc03f8000001e00f803f0fffffe9f871fe01f8000001f01f807f3fffffebf8f3fe01f8000001f83f847f3fffffebfc43ff01f8000003f83f803f3fffffebfc007f05f0000203f81dc43f3fffffe7fe047f04f0000203f81dcc3f0fffffe7ff0c7f04e0000007fc19c43f0fffffe3ff0cff00fc000007fc31cc3f2fffffe3f909ff00fc000007fc31c43f3fffffe3f819ff80fc000007fc39c43f3fffffe7f81bff80f8000007fc7fe43f3fffffe7fc33fe0070000007fc5fe03c0fffffeffc37fe127000000ffc1fe03c8fffffeffc37fc1020000007fe1cc03c9fffffe7fe21fc0000000007fe1cc03cdfffffe7fe20fc000000000ffe1fc43ccfffffe7fe38fc000000000ff81fcc3c8dffffe7f878f8000000000ff81fccbc80ffffe7f0fc78000000000ff80fccbe80ffffe7f1fc780000000007fc0fcc9e88ffffe7f1fc380000000007fc0fcc3ef9ffffe7f1fc300000000007fe07cc7e5fffffe7e3fc30000000000ffe07cc7e7fffffebe3fc30000000000ffe078c3e7fffffebc3fc30000000000fff038c7e7fffffebc3fc30000000000fff038c7e3fffffebc7fc300000000005f801887f3fffffeb87fc300000000001f80100ff3fffffeb0ffc300000000001f80100ff3fffffeb0ffc300000000001f80100ff3fffffea0ffc300000000001f80001ff2fffffe80ffc300000000001f00001ff0fffffecbffe000000000000f00000ff3fffffecdffe000000000000700004ff3fffffeffffe000000000000701004ff3fffffefffff40180000000060000cff3fffffefffffc03f0000000060001cff2fffffeeffffc03c0000000020003cff0fffffef3fff80fe000000000000fcff3fffffef8fff07ffe00000000001fcff3fffffefc3fc1fffe00000000007fcff3fffffeff0003fff40000001fffffcff3fffffefffdffff800000001fffffdff2fffffeffffffffc00000001fffffdff0fffffeffffffff800000001dffffdff3fffffefffffffff82000001dffffdff3fffffefffffffffc6000001fffffdff3fffffefffffffffc2000001fffffdff2fffffefffffffffc2000003fffffdff2fffffeffffffffe8000000400fffdff0fffffefe003ffff000000000007fdff3fffffef07c07ffd000000000001fdff3fffffee7ff81ffc8000000000007dff3fffffeffffe07ff8000000000003cff3fffffefffff03fe0000000000001cff2fffffefffffc1fe0000000000040cff07ffffefffffc0f400000000000c04ff3fffffefffffe070000000000000c1ff3fffffefffff7030000000000008e1ff1fffffefffffb83000000000000061ff3fffffefffff3800000000000000e1ff0fffffef9fffb800000000000000f0ff00ffffef8fff3000000000000001f0ff38ffffef07ff0000000000000003f07f3affffef67ff0000000000000003f87f38ffffee73fa0000000000000007f87f3affffeef9f08000000000006007fc7f08ffffeef9f1c00000000000700ff8bf00ffffecf9e3e00000000000f80ff8bf08ffffecfce7e00000000000fc0ff83f0bffffedfce7e00000000001fc0ff83f0bffffecfce7e00000000003fe0ffc3e09ffffecfc64e00000000003fe1ffc1e01ffffecfe4cc08000000007ff1ffc1e01ffffec7c4cc08000000007ff1f3c1c01ffffecfc0fc08000000007fc1f3c1001ffffe8fc0fc08000000007fc2f3c1001ffffeefc8781dfffff3707fe07365f0bffffeefcc781dfffff3787fe06201f07ffffee7ec701dfffff37c7fe06f07f01ffffee3ec701dfffff37c7fe0ff07f00ffffec7ccf03dfffff37c7fe0ff0ff02ffffee7cef0fdbffff33c7fe0ff07f02ffffef7ce79ffbffff33c7fe0ff87f02ffffef78e73ffbffff13e7fc0ff87f02ffffef79e73fbbffff13e7fc0ff8ff00ffffef79f33fbbffff13e7f807f8ff00ffffef3bf33ff3ffff1163f803e0ff00ffffefb3fb7f77ffff09e3f803e07d00ffffefb3f87f27ffff09f1f001c06900ffffefa7fc7fa7ffff08b0f081c04000ffffef87fcfe67ffff44e0f101c04100ffffefcffffe4fffff0450f001c08000ffffeffffffccfffff00607001800000ffffeffffffc8ffffe00306100800000ffffeffffff99ffffe00106000800000ffffe000000000000000000000000000ffffe000000000000000000000000000ffffffffffe23a000200000000000000ffffffffffc476000200000000000000ffffffffff8872000200000000000000fffffbffff10e4000200001000000000fffffdbffc0000000200000000000000fffffe03800300000200000000000000ffffff00000200000200000000000000ffffffc0000000000200000000000000fffffff8000000000200000000000000fffffffff08000000200000000000000fffffffff80000000200000000000000"`;

exports[`Mask shall overlay the mask on the image 1`] = `"fffffffffffffff7800000000001ff03ffffffffffffffe7800000000003ff03ffffffffffffffffa00000000000ff03ffffffffffffffffe00000000000ff0bfffffffff7ffffffe000000000007f0bffffffffffffffffc000000000007f03fffffffffffffffff000000000183f03fffffffffffffffff0000000000e3f03ffffffffffdfffffc000000000071f0bffffffffffefffffe000000000038f0bffffffffffe7ffffe00000000001cf0bfffffffffff3ffff800000000001ff0bfffffffffff9ffffe00000000000bf07fffffffffffc7fffe00000000000ff0fffffffc7fff07fffe000000000007f3fffffff43ff803fffe000000000007f3fffffff01ff0037ffa000000000107f3ffffffe00ff0017ffc000000000003f3ffffffe007e0013ff8000000000083f2ffffffc007c0003ffc0000000000c7f0ffffffc0078000bffe0000000000e7f3ffffff800f0000fffe0000000000e7f3ffffff800600005ffe0000000000e7f3ffffff800000005fff0020000000f7f3ffffff0000e0005ffc0020000000f3f2fffffe0000e0006ffe0000000000f3f0fffffe0000c0004ffe0000000000fbf3ffffff8001c0002fff0000000000f3f3ffffff800180002ffe0000000000fbf3ffffff000380002ffe00000000007bf3ffffff000380002ffe00000000007fc0fffffe00030001affc00000000003fcdfffffe00030001a7e000000000003bcfffffff00030001a7e0000000000023cfffffff00030000a7e00000000000ffccffffff00038000a7e00000000000ffe8dfffff8007c00007e80000000000ffe88fffff000fc00007c00010000000ffea8fffff001fc00007a00010000000ffed8fffff801fc00007800000000000ffefdfffff803fe0000f800000000000ffe7ffffff803fe0000f800000000000ffe7ffffff803fc0000f000000000000ffe7ffffff807fc0000fa00000000000efe7ffffff807fe0000f800000000000efe7ffffff807fc00017400000000001d7f3ffffff807fe0001f800000000000dff3ffffff80ffc0001f000000000001aff3ffffff80ffc0001f8000000000011ff3ffffffc0ffe0003f8000000002001ff2ffffffc9ffe0007f000000000e003ff2ffffffcfffe0007d0000000007003ff3ffffffddffe000fe0000000007807ff3fffffffffff401fec000000007c0fff3fffffffffff403ff000000000000fff3fffffffffffe07ff000000000001fff2fffffffffffc1fff800000000003fff0fffffff3fff8bfff80000100000ffff3fffffff9fff1ffffe0000000001ffff3fffffffcffc7ffffc80013f000fffff3ffffffff800ffffff80003fffffffff3fffffffffffffffff80001fffffffff2fffffffffffffffff40001fffffffff0fffffffffffffffffc0001fffffffff3fffffffffffffffffd00003ffffffff3fffffffffffffffff69000fffffffff3fffffffffffffffffc1000fffffffff2ffffffffffffffffff1000fffffffff2fffffffffffffffffc8001e187fffff0fffffffc003ffffff7800380007ffff3fffffff0ff07fffff6000000001ffff3ffffffefffc1fffffc0000000007fff3fffffffffff0fffff42000000003fff3fffffffffff83ffef40000000001fff2fffffffffffc1ffffc00002000e0fff0fffffffffffe1ffff000004000c07ff3fffffffffffe0ffff000008001c03ff3ffffffffffff17ffe808000000803ff3ffffffffffff8bfffc08010000001ff3ffffffffffff89f4e008000000001ff0fffffffbffffc4f0e008000000000ff00ffffff8fffbc4f08008000000000ff38ffffff07ff8002440080000000007f3affffff07ff8024000080000000007f3affffff03fe8000000080000000007f3afffffe03f0000000000000000000ff2afffffe01f0001000000000000000bf00fffffc01e0001000000000000000bf08fffffc00e0000000000000000000bf0bfffffc00e00000000000000000003f0bfffffc00e00088000000000000003f0bfffffc00e00088000000000000001f03fffffc00600088000000000000001f03fffffc00c00080000000000000001f0bfffffc00800080000000000000001e01fffffc00000080000000000000001e01ffffff00d801dfffff77800000007f0bffffff00e803dfffff77800000007f07fffffe00c003dfffff77c00000007f01fffffe00c007dfffff77c00000037f00fffffe00c007dfffff37c0000003ff02fffffe00e00fffffff33c0000003ff0affffff00e03fffffff33e0000000ff0affffff00e03ffbffff33e0000000ff02ffffff01e03fbbffff13e0000000ff00ffffff01f03fffffff13e0000000ff00ffffff03f83ff7ffff11e0000000ff00ffffff83f87f77ffff19f0000000ff01ffffff87f87f37ffff49f00040007d01ffffff87fc7fe7ffff48f00080007301ffffffcffcfe6fffff44f801c000c100ffffffcfffff4fffff447801c0008100fffffffffffccfffff427801a0008000fffffffffffe8ffffe42380100000000fffffffffff99ffffe40180000000000fffffffffff13ffffe00180000000000fffffffffff23ffffe00080000000000ffffffffffe67ffffe00000000000000ffffffffffc47fdefe00000000000000ffffffffff88ffc5fe00001000000000fffffbffff10fe0e0e00001800000000fffffdfffc21ef000600000000000000fffffe030003dc000200000000000000ffffff80000fbc000600000000000000ffffffe0001f80000600000000000000ffffffff007e00000200000000000000fffffffffffd00000200000000000000ffffffffffe000000200000000000000"`;

exports[`Mask shall overlay the mask on the image 2`] = `"fffffffffffffff4000000000000ffb3ffffffffffffffe40000000000007f5bfffffffffffffffc0000000000007f43fffffffffffffffe0000000000003f03ffffffffe7fffffe0000000000001f0bfffffffffdfffffe0000000000001f53ffffffffff7fffff0000000000003f13ffffffffffbfffff0000000000001f1bffffffffffdfffff8000000000041f8bffffffffffefffff0000000000030f4bffffffffffe7ffff0000000000018f0bfffffffffff3ffff000000000001070bfffffffffff99ffe0000000000009713fffffffffffc5fff0000000002008f0bffffffc7fff46fff0000010006007f3bfffffff3ffbc2ffe0000018009007f3fffffffc9ffec27ff0000014011007f3fffffff00fe8407ff0000012011803f3fffffff067d8603fe0000011010403fbffffffe007a0303fe0000020020483f1ffffffc027e0103ff8000020800287fbffffffc04f40089ff80000404404c7f3ffffffc02680085ff00000400402e3f3ffffffc01078085ff00020400203f7fbffffff8010e8085fe00020000201e3fbffffff0008f0004ff00000802002f3f9ffffffc008f0084ff80000800002f3fbffffffc060e0040ff00000000002f3fbffffff8059c0040ff8000000000173fbffffff8023c0020ff000000000017bf3ffffff8023c00c2ff0000000000173c0ffffff00038011aff00001001e00bbccffffff00038001a7000000001000b3cfffffff80138001a6800000000003a3cfffffff8003e000a7000001000002efcfffffff8003900027000001006800fbcadfffff807fd00007800000007200ffef8fffff800fd04006800001002002ffef8fffff801fc80006000000002002ffef8fffff809fc80006000000001102ffefdfffff80bfe40006000000001002f7e7ffffff803fc40002000001000002f7e7ffffff813fc40000000001000002ebe7ffffff813fc40002000001000800efe7ffffffc27fc40000000001000804cfe7ffffffc27fc40000000001007804d7f3ffffffc47fc48008000000c040048ff3ffffffc4ffc400000000002040088ffbffffffc8ffc400020000000040080ffbffffffd0ffe480020000002040281ffbffffffd9ffe4806c000000200c103ff2ffffffffffe4807c0000001082103ff3fffffffdffe000fa0000000083907ffbfffffffffff701f8000000008140fff3fffffffffff701fc000000000000fff3ffffffffffff03fe000000090001fffbffffffeffffc07fe000000090003fff9fffffff3fff81ffe00000002000ffffbfffffff9fff1ffff80000000001ffff3fffffffc3fc3ffff8000007000fffffbffffffff800fffffe00000fffffffff3fffffffffffffffff000007ffffffffaffffffffffffffffe000007ffffffff9fffffffffffffffd300000fffffffffbffffffffffffffffe400003ffffffffbfffffffffffffffde000003ffffffffbffffffffffffffffe400007ffffffffbfffffffffffffffdf400007ffffffffbfffffffffffffffc3c0000e007fffff8fffffffc003fffffe0000100007ffff3fffffff0fe07fffef0000000001ffffbffffffefffc1fffff00000000007fffbfffffffffff07ffaf00000000003fffbfffffffffff83ff8600000000019fff3fffffffffffc1ffce000000000e8fff1fffffffffffe0ffde000000000ce7ff3fffffffffffe0fe82000000000033ff3ffffffffffff07e84000000000893ff3ffffffffffff83e04000000000011ff3fffffffffffb8100c008000000111ff8fffffffffffbc4e00008000000208ff88ffffffefffb40200008000000008ff38ffffff97ff800000008000c004047fbaffffff9fff802000008000c008047f3affffff8bffc00000000000b008027f3affffff07f2600000000000881002ff28ffffff05f4201000000000041002bf00fffffd05ec100000000001020000bf08ffffff02f80000000000020200003f0bfffffe02f00000000000020100023f0bfffffd02e81080000000040100023f0bfffffd01e80080000000080080021f13fffffd014800880000000800a0021f0bfffffd02d00080000000000080001f01fffffd02102080000000080100001e01fffff902302080000000080000011001ffffff02b841dfffff77080130097f0bffffff02d841dfffff37800100057f1fffffff82d843dfffff37c001000b7f19ffffffc2c843dfffff37c00110097f00ffffff82c043dfffff37d0011007ff42ffffff82e0efdfffff33c80100057f42ffffff86e85ffbffff33c80100047f02ffffff84e8fffbffff13e80300047f5affffff85e83fbbffff13e8020004ff00ffffff85fcbfbbffff13e8040004ff00ffffffcbfcbff3ffff11ec041414ff00ffffffcbfcff77ffff09e4040018fd01ffffffcffb7f37ffff09f20400104900ffffffffff7fe7ffff08b10080204108fffffffffffe67ffff44e10980204100fffffffffffe4fffff44710882208100fffffffffffccffffe00710902200000fffffffffffc8ffffe00308900400000fffffffffff99ffffe00109001400000fffffffffff11fbf3e00186001800008fffffffffff23f8fbe00006000800000ffffffffffe63e802600000000000000ffffffffffc47e002600000000000000fffff7ffff8877000200000000000000fffffbffff10f4000200001000000000fffffdbffc0004000200000000000008fffffe01000300000200000000000000ffffff80000700000200000000000000ffffffe0000800000200000000000000fffffffe002000000200000000000000ffffffffff8000000200000000000048fffffffff80000000200000000000008"`;

exports[`Mask shall return the connected components 1`] = `"ffffffffffffffe7800000000001ff03ffffffffffffffe7800000000001ff03ffffffffffffffff800000000000ff03ffffffffffffffffc00000000000ff0bfffffffff7ffffffc000000000007f0bffffffffffffffffc000000000007f03ffffffffffffffffd000000000183f03ffffffffffffffffd000000000063f03ffffffffffdfffffc000000000071f0bffffffffffefffffe000000000039f0bffffffffffe7ffffe00000000001cf0bfffffffffff3ffff800000000001ff0bfffffffffff9ffffc000000000009f03fffffffffffc7fffe00000000000ff0fffffff87fff07fffe000000000007f3fffffff03ffb03fffe000000000007f3fffffff01ffc037ffa000000000107f3ffffffe00ff8017ffc000000000003f3ffffffc007f8613ff8000000000083f0ffffffc007a0003ff80000000000c7f0ffffff8007c010bffe0000000000e7f3ffffff800fc008dffe0000000000e7f3ffffff800680085ffe0000000000e7f3ffffff800060005ffe0020000000f7f3ffffff0000e8005ffc0020000001e3f2fffffe0000f0007ffe0000000000f3f0fffffe0000f0004ffe0000000000fbf2fffffe8000e0002ffe0000000000f3f3ffffff0001c0042ffe0000000000fbf3ffffff000380002ffe00000000007bf3ffffff0003800c2ffc00000000007fc0fffffe00038001affc00000000003fccfffffe00038001a7f000000000003bcffffffe00020001a7e000000000003bcfffffff0002e000a7f00000000000ffccffffff00038000a7e00000000000ffe8dfffff0007d00007e80000000000ffe88fffff000fd00007c00010000000ffe88fffff001fc80007a00010000000ffe88fffff001fc80007800000000000ffefdfffff003fe0000f800000000000ffefffffff003fe4000fc00000000000ffe7ffffff803fe0000f800000000000ffe7ffffff803fc00007800000000000efe7ffffff807fe0000fc00000000000efe7ffffff807fc08017c00000000001dff3ffffff807fe0001f800000000000dff3ffffff80ffc0001f8000000000018ff3ffffff80ffc0003f8000000000011ff3ffffff80ffe4003f8000000002001ff2ffffff89ffe4007f800000000e003ff0ffffffcfffe0007d8000000003003ff3ffffffcdffe000ff0000000007807ff3fffffffffff401ffc000000007c0fff3fffffffffff703ff000000000000fff3fffffffffffc07fe000000000001fff2fffffffffffe1fff800000000003fff0fffffff3fff83fff00000000000ffff3fffffff9fff1ffffc0000000001ffff3fffffffdffc7fffff8041af000fffff3ffffffff800ffffff80403fffffffff3fffffffffffffffff80001fffffffff2fffffffffffffffff80001fffffffff0fffffffffffffffffc0000fffffffff3ffffffffffffffffff00007ffffffff3fffffffffffffffff68000fffffffff3fffffffffffffffffd8000fffffffff2fffffffffffffffffe8000fffffffff2fffffffffffffffffc8001e19ffffff0fffffffe007fffffff800380007ffff3fffffff0fe07fffffe000400001ffff3ffffffefffc1fffffc0000000007fff3ffffffffffe0fffff40000000003fff3fffffffffff83ffff60000000001fff2fffffffffffc1ffffc00000000e0fff0fffffffffffe1ffff400004000c07ff3fffffffffffe0ffff000000000403ff3ffffffffffff17ffe808000000803ff3ffffffffffff83fff808000000001ff3ffffffffffff89f5e008000000001ff0fffffff9ffffc4e4e008000000000ff00ffffff8fffbc4f0e008000000000ff38ffffff07ff00064c008000000000ff3affffff07ff0024000080000000007f3afffffe03fe0004000080000000007f3afffffe03f8000000008000000000ff2afffffe01f0001000000000000000bf00fffffc01f0001000000000000000bf08fffffc01e0000000000000000000bf0bfffffc00e00000000000000000003f0bfffffc00e00080000000000000003f0bfffffc00e00088000000000000001f03fffffc00600088000000000000001f03fffffc00c00088000000000000001f03fffffc00800080000000000000001e01fffffc00000080000000000000001e01fffffe00c001dfffff77800000007f0bfffffe00c003dfffff77800000007f0ffffffe00c003dfffff77c00000007f01fffffe00c007dfffff77c0000001ff08fffffc00c007dfffff37c0000003ff02fffffe00e00fffffff33c0000003ff0affffff00e01fffffff33e0000001ff0affffff00e03ffbffff33e0000000ff02ffffff01e03fbbffff13e0000000ff00ffffff01f03fffffff13e0000000ff00ffffff03f83ff7ffff11e0000000ff00ffffff83f87f77ffff19f0000000ff01ffffff87f87f37ffff49f000c0007f01ffffff87fc7fe7ffff48f00080007b01ffffffcffcfeefffff44f801c000c100ffffffcfffff6fffff447801c0008100fffffffffffdcfffff40780180008100fffffffffffc8ffffe40380180000000fffffffffff99ffffe40180040000000fffffffffff11ffffe20180000000000fffffffffff33ffffe00080000000000ffffffffffe67ffffe00000000000000ffffffffffc47fdffe00000000000000ffffffffff88ffcffe00000000000000fffffbffff10ff062e00001800000000fffffdfffc21ff000e00000000000000fffffe438003ff000200000000000000ffffff800007bc000600000000000000ffffffe0001f88000600000000000000fffffffe007f00000200000000000000fffffffffffd00000200000000000000fffffffffff000000200000000000000"`;
